pkg/adapters/rules/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,284 @@
+
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			scheme.AddKnownTypes(
+				groupversion,
+				&RoutingRule{},
+				&RoutingRuleList{},
+				&api.ListOptions{},
+				&api.DeleteOptions{},
+			)
+			return nil
+		})
+	schemeBuilder.AddToScheme(api.Scheme)
+
+	client, err := rest.RESTClientFor(kubeConfig)
+	if err != nil {
+		logger.Printf("Failed to create the in-cluster client. %s\n", err)
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// Stop synchronizing the Kubernetes adapter.
+func (a *Adapter) Stop() error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	if a.stopChan == nil {
+		err := fmt.Errorf("kubernetes adapter not started")
+		logger.WithError(err).Errorf("Failed stopping Kubernetes adapter")
+		return err
+	}
+	close(a.stopChan)
+	a.stopChan = nil
+	a.workqueue.Stop()
+	return nil
+}
+
+// ListRules queries for the list of rules currently exist.
+func (a *Adapter) ListRules(f *a8api.RuleFilter) (*a8api.RulesSet, error) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	rules := make([]a8api.Rule, 0, len(a.rules))
+	for _, rule := range a.rules {
+		rules = append(rules, *rule)
+	}
+	rules = f.Apply(rules)
+	rulesSet := &a8api.RulesSet{
+		Rules:    rules,
+		Revision: 1, // TODO: how to fill revision ?
+
+	}
+	logger.Info("List rules: %v", rulesSet.Rules )
+	return rulesSet, nil
+}
+
+// addRule is the callback invoked by the Kubernetes cache when a rule API resource is added.
+func (a *Adapter) addRule(obj interface{}) {
+	rule, ok := obj.(*RoutingRule)
+	if !ok {
+		logger.Warnf("Invalid rule added: object is of type %T", obj)
+		return
+	}
+	if rule.Status.State == RuleStateInvalid {
+		logger.Warnf("Rule state is invalid. Rule %#v will not be added", rule)
+		return
+	} else if rule.Status.State == RuleStateValid {
+		a.revision++
+		a8Rule := &rule.Spec
+		ruleName := rule.Metadata.Name
+		a.rules[ruleName] = a8Rule
+		return
+	} else {
+		logger.Warnf("Rule state %s is undefiend . Rule %#v will not be added", rule.Status.State, rule)
+		return
+	}
+}
+
+// updateRule is the callback invoked by the Kubernetes cache when a rule API resource is updated.
+func (a *Adapter) updateRules(oldObj, newObj interface{}) {
+	oldRule, ok := oldObj.(*RoutingRule)
+	if !ok {
+		logger.Warnf("Invalid rule update: old object is of type %T\n", oldObj)
+		return
+	}
+	newRule, ok := newObj.(*RoutingRule)
+	if !ok {
+		logger.Warnf("Invalid rule update: new object is of type %T\n", newObj)
+		return
+	}
+
+	if newRule.Status.State == RuleStateInvalid {
+		a.revision++
+		delete(a.rules, oldRule.Metadata.Name)
+		return
+	} else if newRule.Status.State == RuleStateValid {
+		a.revision++
+		a8Rule := &newRule.Spec
+		ruleName := newRule.Metadata.Name
+		a.rules[ruleName] = a8Rule
+		return
+	} else {
+		logger.Warnf("Updated rule state %s is undefiend . Rule %#v will not be updated", newRule.Status.State, oldRule)
+		return
+	}
+}
+
+func (a *Adapter) deleteRules(obj interface{}) {
```

Comment:
Note that `obj` here might not be the actual `RoutingRule` object, but a wrapper around it.
There's an `extractDeletedObject()` function in the [Kubernetes discovery adapter](https://github.com/amalgam8/amalgam8/blob/master/pkg/adapters/discovery/kubernetes/kubernetes.go#L495) that handles this; Let's move this function to the `amalgam8/pkg/kubernetes` package, and re-use it from here.